### DISCUSSION TITLE
How can I make Direct/StyledProperty not set DataContext?

### DISCUSSION THREAD

**Discussion by Phoenixkaze** [AvaloniaUI/Avalonia#14236] [Q&A] [answered]
I have a custom control with a view model as DataContext. 

The control is embedded in a ItemsRepeater. The item will be **part of the view model**.

 I defined a DirectProperty on it for binding the item in xaml.

But I found after setting the property the data context had become the value of the property, which is not what I want.

Is there any way I can prevent this behavior? Or what's the proper way to do this?

**Comment by maxkatz6** [comment]
I am not sure if I understood what exactly you want to do. But it is by design of DataContext to be inherited by rules of the visual tree. Items control like ItemsRepeater will assign item object as a DataContext of each child - that's how it supposed to work. And it's not possible to override this behavior.
Depending on a task, it might be better to leave DataContext to do its job - inherit collection item, and add another property for your another view model.

**Reply by Phoenixkaze** [reply]
Essentially what I'm doing is like this:

In a user control I did something like this:
```xaml
<ItemsRepeater ItemsSource="{Binding Posts}">
  <ItemsRepeater.Layout>
    <UniformGridLayout >
  </ItemsRepeater.Layout>
  <ItemsRepeater.ItemTemplate>
    <DataTemplate>
      <views:PostCell Post="{Binding}"/>
    </DataTemplate>
  </ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```
where `Posts` is a collection of posts.

The core of `PostCell` is an `Image`. I've got image urls in `Post`.

`PostCellViewModel` is the view model ( and DataContext ) of `PostCell`.  Its only property is the post for now. It depends on the post bound to the post cell.

Since I have defined a `Post` direct property on `PostCell`，binding it causes change to `DataContext`.

What I want is: Pass a post into `PostCell` in xaml binding ( then set the post property of view model or possibly construct a view model directly ) and still use the view model as `DataContext`.

**Reply by timunie** [reply]
I think it needs DataType decined for the DataTemplate

**Reply by Phoenixkaze** [reply]
I set DataType on UserControl level. The post cell in data template can get the post as expected. I just...need the bound post but do not want it to be the cell's DataContext.

**Reply by timunie** [reply]
Doesn't make much sense to me? It needs to know the DataContext for the Binding to work. You can use a different Source or a Converter probably

**Reply by Phoenixkaze** [reply]
Because I have a `PostCellViewModel` class as the cell's `DataContext`. The post should be part of the view model (the view model has a post property), not the whole data context. The xaml of `PostCell` has some bindings on the view model, not directly on the post. Or maybe I misused `DataContext`？

**Reply by timunie** [reply]
looks like you did. Can you prepare a very minimal sample and upload it here?

**Reply by Phoenixkaze** [reply]
Here it is [Sample.zip](https://github.com/AvaloniaUI/Avalonia/files/13971368/Sample.zip)

Sorry if I've made you confused. I guess I need to clarify my question. 

While the binding seems **working**, I keep getting `InvalidCastException` ( from `Post` to `PostCellViewModel` ) when debugging the app. Though it doesn't break the app by default. And I got to know that the data context had been set to `Post`. So I feel confused by this.

Is this the normal behavior? If the control is binding a `PostCellViewModel` why it then changes to a `Post`? Maybe I misunderstand how data context working.

**Reply by timunie** [reply, marked as answer]
Your idea doesn't really fly here. Your control shouldn't create it's own ViewModel. What would work is to add PostViewModel as a wrapper around your items. Another idea would be to create a TemplatedControl or use a converter to get the image shown.

**Reply by timunie** [reply]
Added a fix which also supports async image loading. 
[Sample_Fixed.zip](https://github.com/AvaloniaUI/Avalonia/files/13974629/Sample_Fixed.zip)

**Reply by Phoenixkaze** [reply]
I see. Thank you for help and you time!
